### PR TITLE
Query, Header Default Value Tag Feature

### PR DESCRIPTION
**Pull request by muratmirgun**
## Description

This Feature provides an approach in Go to assign default values to the fields of structs. The tagHandlers function assigns these default values based on the specific type of a field. To enhance performance, we cache which fields of the structs have default values (structCache), so we don't have to repeatedly look up these fields every time. Because Handlers Query structs not change

Fixes:
#2571 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [ ] For new functionalities I follow the inspiration of the express js framework and built them similar in usage
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation - /docs/ directory for https://docs.gofiber.io/
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] If new dependencies exist, I have checked that they are really necessary and agreed with the maintainers/community (we want to have as few dependencies as possible)
- [x] I tried to make my code as fast as possible with as few allocations as possible
- [x] For new code I have written benchmarks so that they can be analyzed and improved

## Commit formatting:

Use emojis on commit messages so it provides an easy way of identifying the purpose or intention of a commit. Check out the emoji cheatsheet here: https://gitmoji.carloscuesta.me/

### CODE REVIEW

**Comment by ReneWerner87** [comment]
Thx

Can you share the benchmark results
How big is the impact for cases without defaults, do we have pay for that as well?

Can you put the default value handling  and the tests in a separate go file (don't like big files, I'm more of a code divider)?
If we have this built in for queryparser, it would probably make sense to build it in for the other parsers (body & header) as well.

Should we add a config setting so you can disable this handling to get the maximum performance if you don't need defaults or have your own handling?

**Comment by muratmirgun** [comment]
> Thx
> 
> Can you share the benchmark results How big is the impact for cases without defaults, do we have pay for that as well?
> 
> Can you put the default value handling and the tests in a separate go file (don't like big files, I'm more of a code divider)? If we have this built in for queryparser, it would probably make sense to build it in for the other parsers (body & header) as well.
> 
> Should we add a config setting so you can disable this handling to get the maximum performance if you don't need defaults or have your own handling?

First, let me share the benchmark results. but These examples for ColdStart

Benchmark_Ctx_QueryParser
Benchmark_Ctx_QueryParser-8                                       653192              1621 ns/op             856 B/op         38 allocs/op
Benchmark_Ctx_QueryParser-8                                       751180              1611 ns/op             856 B/op         38 allocs/op
Benchmark_Ctx_QueryParser-8                                       753249              1563 ns/op             856 B/op         38 allocs/op
Benchmark_Ctx_QueryParser-8                                       771404              1646 ns/op             856 B/op         38 allocs/op

Benchmark_Ctx_QueryParserWithDefaultValues
Benchmark_Ctx_QueryParserWithDefaultValues-8             974080              1239 ns/op             720 B/op         25 allocs/op
Benchmark_Ctx_QueryParserWithDefaultValues-8             941806              1245 ns/op             720 B/op         25 allocs/op
Benchmark_Ctx_QueryParserWithDefaultValues-8             963400              1220 ns/op             720 B/op         25 allocs/op
Benchmark_Ctx_QueryParserWithDefaultValues-8             963571              1236 ns/op             720 B/op         25 allocs/op

Benchmarks for multiple Request Scenario:

Benchmark_Ctx_QueryParser
Benchmark_Ctx_QueryParser-8                                        14372             78933 ns/op           42817 B/op       1900 allocs/op
Benchmark_Ctx_QueryParser-8                                        15265             77700 ns/op           42817 B/op       1900 allocs/op
Benchmark_Ctx_QueryParser-8                                        15346             78469 ns/op           42817 B/op       1900 allocs/op
Benchmark_Ctx_QueryParser-8                                        15432             78051 ns/op           42817 B/op       1900 allocs/op

Benchmark_Ctx_QueryParserWithDefaultValues
Benchmark_Ctx_QueryParserWithDefaultValues-8              19756             60628 ns/op           36018 B/op       1250 allocs/op
Benchmark_Ctx_QueryParserWithDefaultValues-8              19581             60628 ns/op           36018 B/op       1250 allocs/op
Benchmark_Ctx_QueryParserWithDefaultValues-8              19682             60993 ns/op           36017 B/op       1250 allocs/op
Benchmark_Ctx_QueryParserWithDefaultValues-8              19599             60733 ns/op           36018 B/op       1250 allocs/op



With multiple

Apart from that, we can do it for all parsers, it will be logical and we can do them with a separate file logic. In the cold start logic, there is a small cost involuntarily at first until the first cache settles, but there will be no problem after the request comes to all the handlers that contain the default structure.

If we are going to integrate it into all parse structures, we can control this with the config for users who are worried about speed.

**Comment by ReneWerner87** [comment]
Can you share the performance without and with your function for the queryparser (i.e. the times from before vs. after the adjustment)

**Comment by muratmirgun** [comment]
> Can you share the performance without and with your function for the queryparser (i.e. the times from before vs. after the adjustment)

Sorry, I forgot it. 

Benchmark_Ctx_QueryParser
Benchmark_Ctx_QueryParser-8               679710              1629 ns/op             856 B/op         38 allocs/op
Benchmark_Ctx_QueryParser-8               740930              1625 ns/op             856 B/op         38 allocs/op
Benchmark_Ctx_QueryParser-8               752588              1584 ns/op             856 B/op         38 allocs/op
Benchmark_Ctx_QueryParser-8               763689              1578 ns/op             856 B/op         38 allocs/op
Benchmark_Ctx_QueryParser_Comma
Benchmark_Ctx_QueryParser_Comma-8         665384              1764 ns/op             928 B/op         44 allocs/op
Benchmark_Ctx_QueryParser_Comma-8         690904              1766 ns/op             928 B/op         44 allocs/op
Benchmark_Ctx_QueryParser_Comma-8         678898              1767 ns/op             928 B/op         44 allocs/op
Benchmark_Ctx_QueryParser_Comma-8         683109              1754 ns/op             928 B/op         44 allocs/op
PASS
ok      github.com/gofiber/fiber/v2     9.871s

**Comment by gaby** [comment]
@muratmirgun Can you take a look at the ci failures

**Comment by ReneWerner87** [comment]
move the functionality to the utils folder

if you can then please add the function for all parsers and add a config setting

**Comment by muratmirgun** [comment]
> move the functionality to the utils folder
> 
> if you can then please add the function for all parsers and add a config setting

Okey I am working on it! 🚀

File: utils/default.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+				field.SetFloat(f)
+			}
+		}
+	case reflect.Bool:
+		if !field.Bool() {
+			if b, err := strconv.ParseBool(tagValue); err == nil {
+				field.SetBool(b)
+			}
+		}
+	case reflect.Slice:
```
Comment (efectn) [inline utils/default.go:34]: Can you add support for single and multiple pointers here?

Comment (muratmirgun) [reply utils/default.go:34]: Okey, I will add it newt commit.

**Comment by ReneWerner87** [comment]
just change the PR type from draft in the normal one, when you are finished with the PR

**Comment by muratmirgun** [comment]
> just change the PR type from draft in the normal one, when you are finished with the PR

Yes I know thanks for remember 💯

File: utils/default.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+					intPtr := reflect.New(elemType.Elem())
+					intPtr.Elem().SetInt(intVal)
+					val = intPtr
+				}
+			}
+		case reflect.String:
+			val = reflect.ValueOf(item)
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			if intVal, err := strconv.ParseInt(item, 10, 64); err == nil {
+				val = reflect.ValueOf(int64(intVal))
```
Comment (efectn) [inline utils/default.go:64]: Do we really need int64() here

Comment (muratmirgun) [reply utils/default.go]: I delete it

**Comment by muratmirgun** [comment]
I saw linter results and I am working on it @efectn

**Review by efectn — APPROVED** [review]
Mostly seems OK to me. Perhaps we can add more features in the future.

**Comment by muratmirgun** [comment]
@efectn I don't have merge access you can merge

**Comment by muratmirgun** [comment]
I have posted the fix code for int slice parser function you can check it. now test codes will not give error

**Comment by ReneWerner87** [comment]
> @efectn I don't have merge access you can merge

Will look again tomorrow and merge then

**Comment by ReneWerner87** [comment]
@muratmirgun can you please look at the lint errors

2. comments
- why do we set the default values after the filling process? it should be easier beforehand and then there is no chance that something will be overwritten by faulty code
- can you add the lines for the `BodyParser`, `ParamParser` and `CookieParser`, it should only be a small block which is added + tests

**Comment by muratmirgun** [comment]
@ReneWerner87 Yes, I will work on the bugs and look at the additions you mentioned

**Comment by muratmirgun** [comment]
@ReneWerner87 Sorry for delay. I've been a little busy because of my job change, so I will update the PR shortly.

**Review by skydiver1980 — APPROVED** [review]

**Comment by ReneWerner87** [comment]
@muratmirgun did you have time to look again?

**Comment by muratmirgun** [comment]
> @muratmirgun did you have time to look again?

Hi, I'll work on it today

File: utils/default.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		if val.IsValid() {
+
+			slice = reflect.Append(slice, val)
+		}
+	}
+
+	field.Set(slice)
+}
+
+var structCache = make(map[reflect.Type][]reflect.StructField)
```
Comment (rohitChaku) [inline utils/default.go:96]: I was wondering if `structCache` here could cause race conditions on concurrent access?Maybe we should add a RW mutex?

**Comment by rohitChaku** [comment]
Hello, I was checking the implementation of `default` values in gin, and it seems pretty neat. I was wondering is we could borrow some constructs from there? I'm primarily referring to the syntax as the logic is pretty similar [form_mapping.go](https://github.com/gin-gonic/gin/blob/master/binding/form_mapping.go#L167C10-L167C10).

I'm relatively new to Golang and open source, so I want to ensure my suggestion aligns with the community's spirit. If there's any concern or if this isn't appropriate, please let me know, and I'll promptly remove my comment. Thanks.

**Comment by ReneWerner87** [comment]
@muratmirgun can you check and consider the last comments, thanks for your work, any help is appreciated

**Comment by ReneWerner87** [comment]
- [ ] pls check the other parsers https://github.com/gofiber/fiber/pull/2699#issuecomment-1805345603
- [ ] can we check https://github.com/gofiber/fiber/pull/2699#discussion_r1422480231 again
- [ ] one last lint error https://github.com/gofiber/fiber/actions/runs/7310291693/job/20152420546?pr=2699

File: utils/default.go (hunk 3 above)
Comment (ReneWerner87) [reply utils/default.go]: fixed with https://github.com/gofiber/fiber/pull/2699/commits/ad89ced1a32fb033133ef1afe3c219a5cfeea702

**Review by ReneWerner87 — APPROVED** [review]

**Comment by muratmirgun** [comment]
Hi, I am sorry for being offline lately. I have had some health problems and I have seen your articles and now I am looking at the example given. @ReneWerner87

**Comment by ReneWerner87** [comment]
> Hi, I am sorry for being offline lately. I have had some health problems and I have seen your articles and now I am looking at the example given. @ReneWerner87

no problem, the pull request should be almost ready i have done the last steps with the other maintainers

health always comes first of course, hope you are well again

**Comment by ReneWerner87** [comment]
image
image

**Comment by muratmirgun** [comment]
> > Hi, I am sorry for being offline lately. I have had some health problems and I have seen your articles and now I am looking at the example given. @ReneWerner87
> 
> no problem, the pull request should be almost ready i have done the last steps with the other maintainers
> 
> health always comes first of course, hope you are well again

thank you very much 🙏 we will move forward again in the next issues🚀

File: ctx.go [hunk 4]
```diff
@@ -410,28 +408,37 @@ func (c *Ctx) BodyParser(out interface{}) error {
...
+			utils.SetDefaultValues(out)
+		}
+
+		switch {
+		case strings.HasPrefix(ctype, MIMETextXML), strings.HasPrefix(ctype, MIMEApplicationXML):
+			if err := xml.Unmarshal(c.Body(), out); err != nil {
+				return fmt.Errorf("failed to unmarshal: %w", err)
+			}
+			return nil
+		case strings.HasSuffix(ctype, "json"):
```
Comment (gaby) [inline ctx.go:437]: Why was `json` moved down here? It used to be the first case

File: utils/default.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+		if val.IsValid() {
+			slice = reflect.Append(slice, val)
+		}
+	}
+
+	field.Set(slice)
+}
+
+var (
+	mu          sync.Mutex
```
Comment (gaby) [inline utils/default.go:97]: Can this be `RWMutex` ?

Comment (efectn) [reply utils/default.go:97]: > Can this be `RWMutex` ?

It can. Then we should not have to be waiting for writing lock

Comment (ReneWerner87) [reply utils/default.go]: yes you are right

**Comment by ReneWerner87** [comment]
@muratmirgun maybe you can help, i found out 2 important things through the review

something is not thread safe -> added parallel test execution and analyzed the problems, unfortunately I can't find the right solution
if i try to correct it i get into a deadlock when running the tests because something is wrong with the mutex hirachy

the defaults cannot be applied deeply, which is what you would expect if you use the function from the utils directly, for example
-> I have created a test that proves this and helps to correct the problem

furthermore, i noticed that the performance could be optimized even further, instead of parsing and converting the value from the default tag every time, you could keep it in the same way or create a slice with clousures, which is later only run through with the element and sets the values
therefore the [go-defaults](https://github.com/mcuadros/go-defaults/tree/master) also used the [filler](https://github.com/mcuadros/go-defaults/blob/master/filler.go)

however, the biggest problem is that the functionality is not thread safe

**Comment by muratmirgun** [comment]
I saw your last commits and new test cases I am looking for your comment maybe we can make a mechanism where we can separate the default parts. I've started working on it. @ReneWerner87

**Comment by muratmirgun** [comment]
I am looking the last changes from this pr for sync

**Comment by ReneWerner87** [comment]
another good library
https://github.com/creasty/defaults/tree/master

https://github.com/gofiber/fiber/issues/2571#issuecomment-1684924938
I'm asking myself again whether we should really integrate this, as the performance could suffer and you can simply use it from outside

```go
import "github.com/creasty/defaults"
```
and
```go
defaults.Set(myStruct)
```
is enough to do it outside

**Comment by muratmirgun** [comment]
that's a good question, I made a structure like this at first with the aim of less dependency because a performance problem on the library side can affect us very seriously. we must be in control. if I can't solve the thread safe problem, I will also examine the structures in this repo on the side. Because

**Comment by ReneWerner87** [comment]
we exclude this feature from the current release for now
if you find a good solution we can put it in the next version or v3

by the way, it is planned to be the last v2 release
but maybe we can still use it as an after feature

**Comment by ReneWerner87** [comment]
@muratmirgun can you work on the last notes? in the v3 branch or should we rather just add documentation and recommend other packages?

**Comment by efectn** [comment]
Hi @muratmirgun the PR is now closed as default feature was added in the last versions of Gorilla schema and it's been just updated for Fiber. Anyway, thanks for your contribution!

**Pull request closed without merging**